src/Data/Singletons/Single/Type.hs

```diff
@@ -32,25 +33,48 @@ singType :: OSet Name      -- the set of bound kind variables in this scope
                 , DKind )  -- the kind of the result type
 singType bound_kvs prom ty = do
   checkVanillaDType ty
-  let (_, cxt, args, res) = unravelVanillaDType ty
-      num_args            = length args
+  let (orig_tvbs, cxt, args, res) = unravelVanillaDType ty
+      num_args                    = length args
   cxt' <- mapM singPred_NC cxt
   arg_names <- replicateM num_args (qNewName "t")
   prom_args <- mapM promoteType_NC args
   prom_res  <- promoteType_NC res
   let args' = map (\n -> singFamily `DAppT` (DVarT n)) arg_names
       res'  = singFamily `DAppT` (foldl apply prom (map DVarT arg_names) `DSigT` prom_res)
       tau   = ravel args' res'
-      -- Make sure to subtract out the bound variables currently in scope, lest we
-      -- accidentally shadow them in this type signature.
-      kv_names_to_bind = foldMap fvDType (prom_args ++ cxt' ++ [prom_res])
-                            OSet.\\ bound_kvs
-      kvs_to_bind      = toList kv_names_to_bind
-  let ty' = DForallT ForallInvis
-                     (map DPlainTV kvs_to_bind ++ zipWith DKindedTV arg_names prom_args) $
-            DConstrainedT cxt' tau
+      kvbs  = singTypeKVBs orig_tvbs prom_args cxt' prom_res
+                           (map DPlainTV (toList bound_kvs))
```

Comment:
I'd prefer to do the `map DPlainTV` in `singTypeKVBs`. Just pass the list of bound names here. Or, even better, the set of names. That's what you have, and indeed the order of these names is irrelevant. That way, the impedance-matching to `deleteFirstsBy` is local in `singTypeKVBs`.

Reply:
`singTypeKVBs` now accepts an `OSet Name` for the set of previously bound names.